qiskit/opflow/operator_base.py

```diff
@@ -119,28 +119,14 @@ def to_matrix(self, massive: bool = False) -> np.ndarray:
         """
         raise NotImplementedError
 
-    @abstractmethod
-    def to_legacy_op(self, massive: bool = False) -> LegacyBaseOperator:
-        r""" Attempt to return the Legacy Operator representation of the Operator. If self is a
-        ``SummedOp`` of ``PauliOps``, will attempt to convert to ``WeightedPauliOperator``,
-        and otherwise will simply convert to ``MatrixOp`` and then to ``MatrixOperator``. The
-        Legacy Operators cannot represent ``StateFns`` or proper ``ListOps`` (meaning not one of
-        the ``ListOp`` subclasses), so an error will be thrown if this method is called on such
-        an Operator. Also, Legacy Operators cannot represent unbound Parameter coeffs, so an error
-        will be thrown if any are present in self.
-
-        Warn if more than 16 qubits to force having to set ``massive=True`` if such a
-        large vector is desired.
+    def to_spmatrix(self) -> spmatrix:
```

Comment:
I am wondering whether this might be preferable, given this is the overall base class, and this is a rather naive implementation, as to whether this would be better as an abstract method to more force any sub-classes to notice the need to implement this, and do so more efficiently when they can, rather than by default getting this memory inefficient naive implementation. ListOp has an implementation, for all ListOps, though its return type differs from this - could this be a default implementation in PrimitiveOp, which are arguably simpler, and leave this here as abstract? The new PauliSumOp,  arguably the most complex 'primitive', already has its own overridden method.

Reply:
Yes, I agree that an abstract method is preferable. But, only `list_op`, `pauli_op`, `pauli_sum_op`, and `dict_state_fn` has `to_spmatrix`. Because the others (`matrix_op`, `circuit_op`, `circuit_state_fn`, vector_state_fn`, etc.) don't have `to_spmatrix`, I think the default implementation is simpler than copying the naive implementation to all such operators.

Reply:
Ok, maybe this can get revisited at some point. I can't help feel that when you need a sparse matrix having to go via a full matrix is just not going to cut it as things get larger.

Reply:
If we change the signature of `NumPyEigensolver._solve(self, operator: OperatorBase)` into `NumPyEigensolver._solve(self, operator: ListOp)` and raise an error for other operator, we can remove `OperatorBase.to_spmatrix`. The current implementation seems to apply only such operators to `NumPyEigensolver`. I'm a bit wondering to add a public method. What do you think?